src/modules/bulletproofs/rangeproof_impl.h

```diff
@@ -0,0 +1,792 @@
+            secp256k1_bulletproof_update_commit(commit, &commitp[i][j], &value_gen[i]);
+        }
+        if (extra_commit != NULL && extra_commit[i] != NULL) {
+            secp256k1_sha256_initialize(&sha256);
+            secp256k1_sha256_write(&sha256, commit, 32);
+            secp256k1_sha256_write(&sha256, extra_commit[i], extra_commit_len[i]);
+            secp256k1_sha256_finalize(&sha256, commit);
+        }
+
+        /* Compute y, z, x */
+        secp256k1_bulletproof_deserialize_point(&age, &proof[i][64], 0, 4);
+        secp256k1_bulletproof_deserialize_point(&sge, &proof[i][64], 1, 4);
+
+        secp256k1_bulletproof_update_commit(commit, &age, &sge);
+        secp256k1_scalar_set_b32(&ecmult_data[i].y, commit, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].y)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+        secp256k1_bulletproof_update_commit(commit, &age, &sge);
+        secp256k1_scalar_set_b32(&ecmult_data[i].z, commit, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].z)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        secp256k1_bulletproof_deserialize_point(&ecmult_data[i].t1, &proof[i][64], 2, 4);
+        secp256k1_bulletproof_deserialize_point(&ecmult_data[i].t2, &proof[i][64], 3, 4);
+
+        secp256k1_bulletproof_update_commit(commit, &ecmult_data[i].t1, &ecmult_data[i].t2);
+        secp256k1_scalar_set_b32(&ecmult_data[i].x, commit, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].x)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+
+        /* compute exponent offsets */
+        secp256k1_scalar_inverse_var(&ecmult_data[i].yinv, &ecmult_data[i].y);  /* TODO somehow batch this w the inner-product argument inverse */
+        ecmult_data[i].yinvn = ecmult_data[i].yinv;
+        for (j = 0; j < secp256k1_floor_lg(nbits); j++) {
+            secp256k1_scalar_sqr(&ecmult_data[i].yinvn, &ecmult_data[i].yinvn);
+        }
+        secp256k1_scalar_sqr(&ecmult_data[i].zsq, &ecmult_data[i].z);
+        secp256k1_scalar_negate(&ecmult_data[i].negz, &ecmult_data[i].z);
+
+        /* Update commit with remaining data for the inner product proof */
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, commit, 32);
+        secp256k1_sha256_write(&sha256, &proof[i][0], 64);
+        secp256k1_sha256_finalize(&sha256, commit);
+
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, commit, 32);
+        secp256k1_sha256_finalize(&sha256, randomizer61);
+        secp256k1_scalar_set_b32(&ecmult_data[i].randomizer61, randomizer61, &overflow);
+        if (overflow || secp256k1_scalar_is_zero(&ecmult_data[i].randomizer61)) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
```

Comment:
With all that hashing in this function, have you considered leaving the hash (context) open and add more bytes it to it whenever there is a new input value?

Reply:
Yeah, my benchmarks show that there's very little speedup to be gained by making the hashes more efficient, so I think it wouldn't be worth the extra complexity and difficulty of reimplementing in a language that didn't have open hash contexts available.

Reply:
That is, if I were maintaining a running sha2 context, somebody later trying to make a python verifier using `hashlib` would have a hard time.

Reply:
FWIW, hashlib has cloneable sha256 state objects :)

```python
>>> import hashlib
>>> x = hashlib.sha256(b"ha")
>>> y = x.copy()
>>> x.update(b"llo")
>>> y.update(b"ll")
>>> y.update(b"o")
>>> x.hexdigest()
'd3751d33f9cd5049c4af2b462735457e4d3baf130bcbb87f389e349fbaeb20b9'
>>> y.hexdigest()
'd3751d33f9cd5049c4af2b462735457e4d3baf130bcbb87f389e349fbaeb20b9'
```